File: tools/f0_archive.py
#!/usr/bin/env python3
"""Export and rebuild the 16-color image archive used by resource F0."""
import argparse
import json
import struct
import zlib
from pathlib import Path

from export_asset import chunk
from generated_files import prune_files
from import_asset import indexed_png


ROM_BASE = 0x08000000
PREFIX = {
    1: "00", 2: "010", 3: "011", 4: "100", 5: "101", 6: "110",
    7: "11100", 8: "11101", 9: "11110", 10: "1111100",
    11: "1111101", 12: "1111110", 13: "111111100",
    14: "111111101", 15: "111111110",
}


class Bits:
    def __init__(self, data):
        self.data = data
        self.offset = 0

    def get(self):
        if self.offset >= len(self.data) * 8:
            raise ValueError("F0 image stream ended before its terminator")
        value = self.data[self.offset // 8] >> (self.offset & 7) & 1
        self.offset += 1
        return value


def selection(bits):
    if bits.get() == 0:
        return 1 if bits.get() == 0 else (2 if bits.get() == 0 else 3)
    if bits.get() == 0:
        return 4 if bits.get() == 0 else 5
    if bits.get() == 0:
        return 6
    if bits.get() == 0:
        return 7 if bits.get() == 0 else 8
    if bits.get() == 0:
        return 9
    if bits.get() == 0:
        return 10 if bits.get() == 0 else 11
    if bits.get() == 0:
        return 12
    if bits.get() == 0:
        return 13 if bits.get() == 0 else 14
    if bits.get() == 0:
        return 15
    return None


def decode_pixels(data):
    bits = Bits(data)
    move_to_front = list(range(16))
    pixels = bytearray()
    while True:
        while bits.get() == 0:
            pixels.append(move_to_front[0])
        index = selection(bits)
        if index is None:
            return bytes(pixels), bits.offset
        value = move_to_front.pop(index)
        move_to_front.insert(0, value)
        pixels.append(value)


def encode_pixels(pixels):
    move_to_front = list(range(16))
    bits = []
    for value in pixels:
        if not 0 <= value < 16:
            raise ValueError("F0 image contains a non-4bpp pixel")
        if value == move_to_front[0]:
            bits.append(0)
            continue
        index = move_to_front.index(value)
        bits.append(1)
        bits.extend(int(bit) for bit in PREFIX[index])
        move_to_front.insert(0, move_to_front.pop(index))
    bits.extend((1,) * 10)
    bits.extend((1,) * (-len(bits) % 8))
    encoded = bytearray(len(bits) // 8)
    for index, value in enumerate(bits):
        encoded[index // 8] |= value << (index & 7)
    return bytes(encoded)


def image_png(pixels, palette):
    if len(pixels) != 32 * 32 or len(palette) != 32:
        raise ValueError("F0 images must be 32x32 with 16 colors")
    colors = bytearray()
    for value in struct.unpack("<16H", palette):
        if value & 0x8000:
            raise ValueError("F0 palette uses an unsupported high bit")
        colors.extend(((value & 31) << 3, (value >> 5 & 31) << 3,
                       (value >> 10 & 31) << 3))
    rows = bytearray()
    for y in range(32):
        rows.append(0)
        row = pixels[y * 32:(y + 1) * 32]
        rows.extend(row[x] << 4 | row[x + 1] for x in range(0, 32, 2))
    return (b"\x89PNG\r\n\x1a\n" +
            chunk(b"IHDR", struct.pack(">IIBBBBB", 32, 32, 4, 3, 0, 0, 0)) +
            chunk(b"PLTE", bytes(colors)) +
            chunk(b"IDAT", zlib.compress(bytes(rows), 9)) +
            chunk(b"IEND", b""))


def read_image(data):
    width, height, pixels, colors = indexed_png(data)
    if (width, height, len(colors)) != (32, 32, 16):
        raise ValueError("F0 source PNG must be 32x32 with 16 colors")
    palette = bytearray()
    for red, green, blue in colors:
        if red & 7 or green & 7 or blue & 7:
            raise ValueError("F0 palette channels must be multiples of eight")
        palette.extend(struct.pack(
            "<H", (red >> 3) | (green >> 3) << 5 | (blue >> 3) << 10))
    return bytes(pixels), bytes(palette)


def preview_png(directory, count, columns=8, scale=2):
    rows = (count + columns - 1) // columns
    width, height = columns * 32 * scale, rows * 32 * scale
    output = bytearray([255]) * (width * height * 3)
    for index in range(count):
        pixels, palette = read_image(
            (directory / f"image_{index:02d}.png").read_bytes())
        colors = []
        for value in struct.unpack("<16H", palette):
            channels = (value & 31, value >> 5 & 31, value >> 10 & 31)
            colors.append(tuple((channel << 3) | (channel >> 2)
                                for channel in channels))
        left, top = index % columns * 32 * scale, index // columns * 32 * scale
        for y in range(32):
            for x in range(32):
                color = bytes(colors[pixels[y * 32 + x]])
                for sy in range(scale):
                    for sx in range(scale):
                        target = (((top + y * scale + sy) * width + left +
                                   x * scale + sx) * 3)
                        output[target:target + 3] = color
    scanlines = b"".join(
        b"\0" + output[y * width * 3:(y + 1) * width * 3]
        for y in range(height))
    return (b"\x89PNG\r\n\x1a\n" +
            chunk(b"IHDR", struct.pack(">IIBBBBB",
                                        width, height, 8, 2, 0, 0, 0)) +
            chunk(b"IDAT", zlib.compress(scanlines, 9)) +
            chunk(b"IEND", b""))


def package(image):
    pixels, palette = read_image(image)
    data = palette + encode_pixels(pixels)
    return data + bytes(-len(data) % 4)


def build_archive(plan, directory):
    count = int(plan["images"])
    packages = [package((directory / f"image_{index:02d}.png").read_bytes())
                for index in range(count)]
    offset = len(plan["entries"]) * 2
    offsets = []
    for payload in packages:
        offsets.append(offset)
        offset += len(payload)
    table = [0 if image is None else offsets[int(image)]
             for image in plan["entries"]]
    if any(value >= 0x10000 for value in table):
        raise ValueError("F0 archive offset exceeds 16 bits")
    return struct.pack(f"<{len(table)}H", *table) + b"".join(packages)


def number(text):
    return int(text, 0)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("rom", type=Path)
    parser.add_argument("--address", type=number, required=True)
    parser.add_argument("--input-end", type=number, required=True)
    parser.add_argument("--plan", type=Path, required=True)
    parser.add_argument("--images", type=Path, required=True)
    parser.add_argument("--preview", type=Path)
    args = parser.parse_args()
    rom = args.rom.read_bytes()
    start, end = args.address - ROM_BASE, args.input_end - ROM_BASE
    archive = rom[start:end]
    first = struct.unpack_from("<H", archive)[0]
    if not first or first % 2 or first > len(archive):
        raise ValueError("F0 archive has no valid offset table")
    entries = struct.unpack_from(f"<{first // 2}H", archive)
    unique = sorted(offset for offset in set(entries) if offset)
    if not unique or unique[0] != first:
        raise ValueError("F0 first image does not follow its offset table")
    bounds = {offset: unique[index + 1] if index + 1 < len(unique)
              else len(archive) for index, offset in enumerate(unique)}
    args.images.mkdir(parents=True, exist_ok=True)
    index_by_offset = {offset: index for index, offset in enumerate(unique)}
    for index, offset in enumerate(unique):
        palette = archive[offset:offset + 32]
        pixels, _ = decode_pixels(archive[offset + 32:bounds[offset]])
        if len(pixels) != 32 * 32:
            raise ValueError("F0 image does not decode to 32x32")
        rebuilt = palette + encode_pixels(pixels)
        rebuilt += bytes(-len(rebuilt) % 4)
        if rebuilt != archive[offset:bounds[offset]]:
            raise ValueError("F0 image package does not re-encode exactly")
        (args.images / f"image_{index:02d}.png").write_bytes(
            image_png(pixels, palette))
    plan = {"format": 1, "codec": "golden-sun-mtf4-archive",
            "images": len(unique),
            "entries": [None if not offset else index_by_offset[offset]
                        for offset in entries]}
    if build_archive(plan, args.images) != archive:
        raise ValueError("rebuilt F0 archive differs")
    prune_files(
        args.images, "image_*.png",
        {f"image_{index:02d}.png" for index in range(len(unique))})
    args.plan.parent.mkdir(parents=True, exist_ok=True)
    args.plan.write_text(json.dumps(plan, separators=(",", ":")) + "\n")
    if args.preview is not None:
        args.preview.parent.mkdir(parents=True, exist_ok=True)
        args.preview.write_bytes(preview_png(args.images, len(unique)))
    print(f"entries={len(entries)} images={len(unique)} encoded={len(archive)}")


if __name__ == "__main__":
    main()

File: tools/f0_archive.ts
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { deflateSync } from "./zlib.ts";
import { chunk } from "./export_asset.ts";
import { prune_files } from "./generated_files.ts";
import { indexed_png } from "./import_asset.ts";

export const ROM_BASE = 0x08000000;
export const PREFIX: Record<number, string> = {
  1: "00", 2: "010", 3: "011", 4: "100", 5: "101", 6: "110",
  7: "11100", 8: "11101", 9: "11110", 10: "1111100", 11: "1111101",
  12: "1111110", 13: "111111100", 14: "111111101", 15: "111111110",
};

export interface F0Plan {
  format?: number;
  codec?: string;
  images: number;
  entries: Array<number | null>;
}

export class Bits {
  data: Uint8Array;
  offset = 0;

  constructor(data: Uint8Array) { this.data = data; }

  get(): number {
    if (this.offset >= this.data.length * 8)
      throw new Error("F0 image stream ended before its terminator");
    return (this.data[this.offset >>> 3] >>> (this.offset++ & 7)) & 1;
  }
}

export function selection(bits: Bits): number | null {
  if (bits.get() === 0) return bits.get() === 0 ? 1 : (bits.get() === 0 ? 2 : 3);
  if (bits.get() === 0) return bits.get() === 0 ? 4 : 5;
  if (bits.get() === 0) return 6;
  if (bits.get() === 0) return bits.get() === 0 ? 7 : 8;
  if (bits.get() === 0) return 9;
  if (bits.get() === 0) return bits.get() === 0 ? 10 : 11;
  if (bits.get() === 0) return 12;
  if (bits.get() === 0) return bits.get() === 0 ? 13 : 14;
  if (bits.get() === 0) return 15;
  return null;
}

export function decode_pixels(data: Uint8Array): [Buffer, number] {
  const bits = new Bits(data);
  const moveToFront = Array.from({ length: 16 }, (_, index) => index);
  const pixels: number[] = [];
  while (true) {
    while (bits.get() === 0) pixels.push(moveToFront[0]);
    const index = selection(bits);
    if (index === null) return [Buffer.from(pixels), bits.offset];
    const [value] = moveToFront.splice(index, 1);
    moveToFront.unshift(value);
    pixels.push(value);
  }
}

export function encode_pixels(pixels: Uint8Array): Buffer {
  const moveToFront = Array.from({ length: 16 }, (_, index) => index);
  const bits: number[] = [];
  for (const value of pixels) {
    if (value < 0 || value >= 16) throw new Error("F0 image contains a non-4bpp pixel");
    if (value === moveToFront[0]) { bits.push(0); continue; }
    const index = moveToFront.indexOf(value);
    bits.push(1, ...Array.from(PREFIX[index], Number));
    const [selected] = moveToFront.splice(index, 1);
    moveToFront.unshift(selected);
  }
  bits.push(...Array(10).fill(1));
  bits.push(...Array((8 - bits.length % 8) % 8).fill(1));
  const encoded = Buffer.alloc(bits.length / 8);
  bits.forEach((value, index) => { encoded[index >>> 3] |= value << (index & 7); });
  return encoded;
}

function pngHeader(width: number, height: number, depth: number, colorType: number): Buffer {
  const header = Buffer.alloc(13);
  header.writeUInt32BE(width, 0); header.writeUInt32BE(height, 4);
  header.set([depth, colorType, 0, 0, 0], 8);
  return header;
}

export function image_png(pixels: Uint8Array, palette: Uint8Array): Buffer {
  if (pixels.length !== 32 * 32 || palette.length !== 32)
    throw new Error("F0 images must be 32x32 with 16 colors");
  const colors = Buffer.alloc(48);
  for (let index = 0; index < 16; index++) {
    const value = palette[index * 2] | (palette[index * 2 + 1] << 8);
    if (value & 0x8000) throw new Error("F0 palette uses an unsupported high bit");
    colors[index * 3] = (value & 31) << 3;
    colors[index * 3 + 1] = ((value >>> 5) & 31) << 3;
    colors[index * 3 + 2] = ((value >>> 10) & 31) << 3;
  }
  const rows = Buffer.alloc(32 * 17);
  let cursor = 0;
  for (let y = 0; y < 32; y++) {
    rows[cursor++] = 0;
    for (let x = 0; x < 32; x += 2)
      rows[cursor++] = (pixels[y * 32 + x] << 4) | pixels[y * 32 + x + 1];
  }
  return Buffer.concat([
    Buffer.from("89504e470d0a1a0a", "hex"), chunk(Buffer.from("IHDR"), pngHeader(32, 32, 4, 3)),
    chunk(Buffer.from("PLTE"), colors), chunk(Buffer.from("IDAT"), deflateSync(rows, { level: 9 })),
    chunk(Buffer.from("IEND"), Buffer.alloc(0)),
  ]);
}

export function read_image(data: Uint8Array): [Buffer, Buffer] {
  const [width, height, pixels, colors] = indexed_png(data);
  if (width !== 32 || height !== 32 || colors.length !== 16)
    throw new Error("F0 source PNG must be 32x32 with 16 colors");
  const palette = Buffer.alloc(32);
  colors.forEach(([red, green, blue], index) => {
    if ((red & 7) || (green & 7) || (blue & 7))
      throw new Error("F0 palette channels must be multiples of eight");
    palette.writeUInt16LE((red >>> 3) | ((green >>> 3) << 5) | ((blue >>> 3) << 10), index * 2);
  });
  return [Buffer.from(pixels), palette];
}

export function preview_png(directory: string, count: number, columns = 8, scale = 2): Buffer {
  const rows = Math.ceil(count / columns);
  const width = columns * 32 * scale;
  const height = rows * 32 * scale;
  const output = Buffer.alloc(width * height * 3, 255);
  for (let index = 0; index < count; index++) {
    const [pixels, palette] = read_image(readFileSync(join(directory, `image_${index.toString().padStart(2, "0")}.png`)));
    const colors: number[][] = [];
    for (let colorIndex = 0; colorIndex < 16; colorIndex++) {
      const value = palette.readUInt16LE(colorIndex * 2);
      colors.push([value & 31, (value >>> 5) & 31, (value >>> 10) & 31].map((channel) => (channel << 3) | (channel >>> 2)));
    }
    const left = (index % columns) * 32 * scale;
    const top = Math.floor(index / columns) * 32 * scale;
    for (let y = 0; y < 32; y++) for (let x = 0; x < 32; x++) {
      const color = colors[pixels[y * 32 + x]];
      for (let sy = 0; sy < scale; sy++) for (let sx = 0; sx < scale; sx++) {
        const target = (((top + y * scale + sy) * width + left + x * scale + sx) * 3);
        output.set(color, target);
      }
    }
  }
  const scanlines = Buffer.alloc(height * (1 + width * 3));
  for (let y = 0; y < height; y++) output.copy(scanlines, y * (1 + width * 3) + 1, y * width * 3, (y + 1) * width * 3);
  return Buffer.concat([
    Buffer.from("89504e470d0a1a0a", "hex"), chunk(Buffer.from("IHDR"), pngHeader(width, height, 8, 2)),
    chunk(Buffer.from("IDAT"), deflateSync(scanlines, { level: 9 })), chunk(Buffer.from("IEND"), Buffer.alloc(0)),
  ]);
}

export function package_image(image: Uint8Array): Buffer {
  const [pixels, palette] = read_image(image);
  const data = Buffer.concat([palette, encode_pixels(pixels)]);
  return Buffer.concat([data, Buffer.alloc((4 - data.length % 4) % 4)]);
}

export function build_archive(plan: F0Plan, directory: string): Buffer {
  const count = Number(plan.images);
  const packages = Array.from({ length: count }, (_, index) =>
    package_image(readFileSync(join(directory, `image_${index.toString().padStart(2, "0")}.png`))));
  let offset = plan.entries.length * 2;
  const offsets: number[] = [];
  for (const payload of packages) { offsets.push(offset); offset += payload.length; }
  const table = plan.entries.map((image) => image === null ? 0 : offsets[Number(image)]);
  if (table.some((value) => value >= 0x10000)) throw new Error("F0 archive offset exceeds 16 bits");
  const header = Buffer.alloc(table.length * 2);
  table.forEach((value, index) => header.writeUInt16LE(value, index * 2));
  return Buffer.concat([header, ...packages]);
}

function value(args: string[], name: string): string | undefined {
  const index = args.indexOf(name);
  return index >= 0 ? args[index + 1] : undefined;
}

function positional(args: string[]): string | undefined {
  const takesValue = new Set(["--address", "--input-end", "--plan", "--images", "--preview"]);
  for (let index = 0; index < args.length; index++) {
    if (takesValue.has(args[index])) { index++; continue; }
    if (!args[index].startsWith("-")) return args[index];
  }
}

function main(args: string[]): void {
  if (args.includes("-h") || args.includes("--help")) {
    console.log("usage: f0_archive.ts ROM --address N --input-end N --plan FILE --images DIR [--preview FILE]");
    return;
  }
  const romPath = positional(args);
  const addressText = value(args, "--address");
  const endText = value(args, "--input-end");
  const planPath = value(args, "--plan");
  const images = value(args, "--images");
  const preview = value(args, "--preview");
  if (!romPath || !addressText || !endText || !planPath || !images)
    throw new Error("rom, --address, --input-end, --plan, and --images are required");
  const rom = readFileSync(romPath);
  const archive = rom.subarray(Number(addressText) - ROM_BASE, Number(endText) - ROM_BASE);
  const first = archive.readUInt16LE(0);
  if (!first || first % 2 || first > archive.length) throw new Error("F0 archive has no valid offset table");
  const entries = Array.from({ length: first / 2 }, (_, index) => archive.readUInt16LE(index * 2));
  const unique = Array.from(new Set(entries.filter(Boolean))).sort((a, b) => a - b);
  if (!unique.length || unique[0] !== first) throw new Error("F0 first image does not follow its offset table");
  mkdirSync(images, { recursive: true });
  const indexByOffset = new Map(unique.map((offset, index) => [offset, index]));
  unique.forEach((offset, index) => {
    const bound = index + 1 < unique.length ? unique[index + 1] : archive.length;
    const palette = archive.subarray(offset, offset + 32);
    const [pixels] = decode_pixels(archive.subarray(offset + 32, bound));
    if (pixels.length !== 32 * 32) throw new Error("F0 image does not decode to 32x32");
    const data = Buffer.concat([palette, encode_pixels(pixels)]);
    const rebuilt = Buffer.concat([data, Buffer.alloc((4 - data.length % 4) % 4)]);
    if (!rebuilt.equals(archive.subarray(offset, bound))) throw new Error("F0 image package does not re-encode exactly");
    writeFileSync(join(images, `image_${index.toString().padStart(2, "0")}.png`), image_png(pixels, palette));
  });
  const plan: F0Plan = {
    format: 1, codec: "golden-sun-mtf4-archive", images: unique.length,
    entries: entries.map((offset) => offset === 0 ? null : indexByOffset.get(offset)!),
  };
  if (!build_archive(plan, images).equals(archive)) throw new Error("rebuilt F0 archive differs");
  prune_files(images, "image_*.png", Array.from({ length: unique.length }, (_, index) => `image_${index.toString().padStart(2, "0")}.png`));
  mkdirSync(dirname(planPath), { recursive: true });
  writeFileSync(planPath, `${JSON.stringify(plan)}\n`);
  if (preview) { mkdirSync(dirname(preview), { recursive: true }); writeFileSync(preview, preview_png(images, unique.length)); }
  console.log(`entries=${entries.length} images=${unique.length} encoded=${archive.length}`);
}

if (import.meta.main) main(process.argv.slice(2));
